### PR TITLE
Adds dev_table_tooling to capture output of allowed commands

### PR DESCRIPTION
**Pull request by seejdev**
This PR addresses #876.

- `kolide_dev_table_tooling` table has a `name`, `args`, and `output` column
- Each platform has a hard-coded set of named allowed commands, which correspond to a strict list of binaries and arguments
- A test exercises the cases where a command name isn't provided, or is unrecognized
- I've supplied some hard-coded allowed commands purely as an example of how they would be specified in the code, these can be replaced with more suitable options, or removed entirely, prior to merging

### CODE REVIEW

File: pkg/osquery/tables/dev_table_tooling/table.go [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/kolide/launcher/pkg/osquery/tables/tablehelpers"
+	"github.com/osquery/osquery-go"
+	"github.com/osquery/osquery-go/plugin/table"
+)
+
+// Encapsulates the binary path(s) of a command allowed to execute
+// along with a strict list of arguments.
+type AllowedCommand struct {
```
Comment (directionless) [inline pkg/osquery/tables/dev_table_tooling/table.go:17]:
```suggestion
// AllowedCommand encapsulates the possible binary path(s) of a command allowed to execute
// along with a strict list of arguments.
type AllowedCommand struct {
```


I've never been sure if structs comments are supposed to start with the struct name (analogous to methods). https://tip.golang.org/doc/comment implies not, but does recommend slightly different phrasing. I'd reword slightly

File: pkg/osquery/tables/dev_table_tooling/table.go [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+	args     []string
+}
+
+type Table struct {
+	client    *osquery.ExtensionManagerClient
+	logger    log.Logger
+	tableName string
+}
+
+func TablePlugin(client *osquery.ExtensionManagerClient, logger log.Logger) *table.Plugin {
```
Comment (directionless) [inline pkg/osquery/tables/dev_table_tooling/table.go:28]: As an FYI we're pretty inconsistent about whether we pass `client` to tables that don't need it. You can keep it, `_` it and not store it in the struct, or not pass it at all. (Some tables do need it, but most do not)

File: pkg/osquery/tables/dev_table_tooling/table.go [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+			level.Info(t.logger).Log("msg", "Command name must not be blank")
+			continue
+		}
+
+		cmd, ok := allowedCommands[name]
+
+		if !ok {
+			level.Info(t.logger).Log("msg", "Command not allowed", "name", name)
+			continue
+		} else {
```
Comment (directionless) [inline pkg/osquery/tables/dev_table_tooling/table.go:58]:
```suggestion
```

With early returns, we can skip the nesting. (deep nesting is hard to read)

File: pkg/osquery/tables/dev_table_tooling/table.go [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+func TablePlugin(client *osquery.ExtensionManagerClient, logger log.Logger) *table.Plugin {
+	columns := []table.ColumnDefinition{
+		table.TextColumn("name"),
+		table.TextColumn("args"),
+		table.TextColumn("output"),
+	}
+
+	t := &Table{
+		client:    client,
+		logger:    logger,
```
Comment (directionless) [inline pkg/osquery/tables/dev_table_tooling/table.go:37]:
```suggestion
		logger:    log.With(logger, "table", "kolide_dev_table_tooling"),
```

Will plumb that through for all subsequent output.

File: pkg/osquery/tables/dev_table_tooling/table.go [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+// along with a strict list of arguments.
+type AllowedCommand struct {
+	binPaths []string
+	args     []string
+}
+
+type Table struct {
+	client    *osquery.ExtensionManagerClient
+	logger    log.Logger
+	tableName string
```
Comment (directionless) [inline pkg/osquery/tables/dev_table_tooling/table.go:25]: You can leave this, or just pass it to NewPlugin.

File: pkg/osquery/tables/dev_table_tooling/table_test.go [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+		commandName    []string
+		expectedResult []map[string]string
+	}{
+		{
+			name: "no command name",
+		},
+		{
+			name:        "malware",
+			commandName: []string{"ransomware.exe"},
+		},
```
Comment (directionless) [inline pkg/osquery/tables/dev_table_tooling/table_test.go:26]: Might be worth adding some kind of happy path testing. Which raises the question of what command. I suspect `echo hello` might be okay, can't quickly double check that's not a shell built on linux, but it should work on mac and windows.

**Review by directionless — COMMENTED** [review]
Table looks really good. A couple of pretty small nits, only thing that I'd really push for is removing the `else` blocks on the early returns.

File: pkg/osquery/tables/dev_table_tooling/table.go (hunk 1 above)
Comment (James-Pickett) [inline pkg/osquery/tables/dev_table_tooling/table.go:17]: I think this can be a un-exported
```suggestion
type allowedCommand struct {
```

File: pkg/osquery/tables/dev_table_tooling/commands_darwin.go [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+
+var allowedCommands = map[string]allowedCommand{
+	"echo": {
+		binPaths: []string{"echo"},
+		args:     []string{"hello"},
+	},
+}
+
+// Platform-specific test data
+var echoHelloOutput = "hello\n"
```
Comment (James-Pickett) [inline pkg/osquery/tables/dev_table_tooling/commands_darwin.go:14]: Feels a little odd to me to have a variable only used for testing outside of a `_test.go` file. I would declare a var in `table_test.go` as a map of os -> output inside the `Test_generate()` like:

```
expectedOutputs := map[string]string{
  "windows": "hello\n",
  "darwin": "hello\n",
  "linux": "hello\r\n",
}
```

then access via

```
expectedOutput := expectedOutputs[runtime.GOOS]
```

inside the test case

**Review by James-Pickett — COMMENTED** [review]
looking good, very close, left comment about test data organization

File: pkg/osquery/tables/dev_table_tooling/commands_darwin.go (hunk 7 above)
Comment (seejdev) [reply pkg/osquery/tables/dev_table_tooling/commands_darwin.go]: Yeah I didn't like the messiness of my first approach - your suggestion is definitely an improvement.

I reworked the test to avoid needing to know about newline characters in [5cc1059e](https://github.com/kolide/launcher/pull/881/commits/5cc1059e4dc0e29ace806ffed8719c8fa6a8589c)

It does add some complexity to the test verification code, but makes the tests easier to specify and maintain.

File: pkg/osquery/tables/dev_table_tooling/table_test.go [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+			got, _ := table.generate(context.Background(), tablehelpers.MockQueryContext(constraints))
+
+			if len(tt.expectedResult) > 0 {
+				assert.Equal(t, tt.expectedResult[0]["name"], got[0]["name"])
+				assert.Equal(t, tt.expectedResult[0]["args"], got[0]["args"])
+
+				// To verify output, let's convert back to utf8
+				decodedOutput, _ := base64.StdEncoding.DecodeString(got[0]["output"])
+				scanner := bufio.NewScanner(bytes.NewReader(decodedOutput))
+				for scanner.Scan() {
```
Comment (James-Pickett) [inline pkg/osquery/tables/dev_table_tooling/table_test.go:56]: that's a neat way to normalize the output

File: pkg/osquery/tables/dev_table_tooling/table_test.go [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+		tt := tt
+		t.Run(tt.name, func(t *testing.T) {
+			t.Parallel()
+
+			constraints := make(map[string][]string)
+			constraints["name"] = tt.commandName
+
+			got, _ := table.generate(context.Background(), tablehelpers.MockQueryContext(constraints))
+
+			if len(tt.expectedResult) > 0 {
```
Comment (James-Pickett) [inline pkg/osquery/tables/dev_table_tooling/table_test.go:49]: let's invert this if statement to reduce nesting

```suggestion
			if len(tt.expectedResult) <= 0 {
			    assert.ElementsMatch(t, tt.expectedResults, got)
			    return
			}
			
			// test for expected results
```

**Review by James-Pickett — DISMISSED** [review]
LGTM! Lets give @directionless a chance to take a look before merging.

File: pkg/osquery/tables/dev_table_tooling/table_test.go [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+	"github.com/kolide/launcher/pkg/osquery/tables/tablehelpers"
+	"github.com/stretchr/testify/assert"
+)
+
+func Test_generate(t *testing.T) {
+	t.Parallel()
+
+	tests := []struct {
+		name           string
+		commandName    []string
```
Comment (directionless) [inline pkg/osquery/tables/dev_table_tooling/table_test.go:20]: Why is this an array? Is it because (in theory) we could pass multiple commands? If so, `commandNames`

File: pkg/osquery/tables/dev_table_tooling/table_test.go [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+
+			constraints := make(map[string][]string)
+			constraints["name"] = tt.commandName
+
+			got, _ := table.generate(context.Background(), tablehelpers.MockQueryContext(constraints))
+
+			if len(tt.expectedResult) <= 0 {
+				assert.ElementsMatch(t, tt.expectedResult, got)
+				return
+			}
```
Comment (directionless) [inline pkg/osquery/tables/dev_table_tooling/table_test.go:52]:
```suggestion
			if len(tt.expectedResult) <= 0 {
				require.Empty(t, got)
				return
			}
```

I think `Empty` conveys the intent. (Though you could leave `assert`)

File: pkg/osquery/tables/dev_table_tooling/table.go [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+		cmd, ok := allowedCommands[name]
+
+		if !ok {
+			level.Info(t.logger).Log("msg", "Command not allowed", "name", name)
+			continue
+		}
+
+		output, err := tablehelpers.Exec(ctx, t.logger, 30, cmd.binPaths, cmd.args)
+		if err != nil {
+			level.Info(t.logger).Log("msg", "dev_table_tooling failed", "name", name, "err", err)
```
Comment (directionless) [inline pkg/osquery/tables/dev_table_tooling/table.go:61]:
```suggestion
			level.Info(t.logger).Log("msg", "execution failed", "name", name, "err", err)
```

**Review by directionless — APPROVED** [review]
I'd take this as-is. I left some smallish nits and clarity suggestions

**Comment by directionless** [comment]
closing and reopening to kick the CLA tool

File: pkg/osquery/tables/dev_table_tooling/table_test.go (hunk 10 above)
Comment (seejdev) [reply pkg/osquery/tables/dev_table_tooling/table_test.go:20]: Good catch, no need for an array here

**Review by directionless — APPROVED** [review]

**Review by James-Pickett — DISMISSED** [review]

**Review by James-Pickett — APPROVED** [review]

**Comment by FritzX6** [comment]
### Wishlist of commands:

**Azure AD binding state:**
```
dsregcmd /status
```

File: pkg/osquery/tables/dev_table_tooling/table.go [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+}
+
+func (t *Table) generate(ctx context.Context, queryContext table.QueryContext) ([]map[string]string, error) {
+	var results []map[string]string
+
+	for _, name := range tablehelpers.GetConstraints(queryContext, "name", tablehelpers.WithDefaults("")) {
+		if name == "" {
+			level.Info(t.logger).Log("msg", "Command name must not be blank")
+			continue
+		}
```
Comment (directionless) [inline pkg/osquery/tables/dev_table_tooling/table.go:48]: This is unlikely to trigger -- if name is nil, we won't enter the forloop. But there's no harm in it.

**Review by directionless — APPROVED** [review]

**Review by James-Pickett — APPROVED** [review]
